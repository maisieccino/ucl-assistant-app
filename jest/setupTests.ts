import '@testing-library/jest-native/extend-expect'
import '@testing-library/react-native/cleanup-after-each'
import 'react-native-gesture-handler/jestSetup'

// Silence the warning: Animated: `useNativeDriver` is not supported because
// the native animated module is missing
// https://github.com/facebook/react-native/issues/11094#issuecomment-263240420
jest.mock(`react-native/Libraries/Animated/src/NativeAnimatedHelper`)
