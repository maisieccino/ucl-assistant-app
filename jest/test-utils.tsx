import { NavigationContainer } from '@react-navigation/native'
import { render as rtlRender, RenderAPI, act } from '@testing-library/react-native'
import React from 'react'
import { Provider } from 'react-redux'
import { Store } from 'redux'
import configureStore from 'redux-mock-store'
import thunk from "redux-thunk"
import debounce from "../lib/debounce"
import { initialState as defaultInitialState } from '../redux/reducers'

const middlewares = [
  debounce.middleware,
  thunk,
]
const mockStore = configureStore(middlewares)

const render = async (
  ui: React.ReactElement,
  {
    initialState = defaultInitialState,
    store = mockStore(initialState) as Store,
    ...renderOptions
  } = {},
): Promise<RenderAPI> => {
  const Wrapper = ({ children }) => (
    <Provider store={store}>
      <NavigationContainer>
        {children}
      </NavigationContainer>
    </Provider>
  )
  const result = rtlRender(ui, { wrapper: Wrapper, ...renderOptions })
  await act(async () => {})
  return result
}

export const waitForEventLoop = async (): Promise<void> => new Promise((resolve) => {
  const done = () => {
    jest.useFakeTimers()
    resolve()
  }
  jest.useRealTimers()
  setTimeout(done, 0)
})

// re-export everything
export * from '@testing-library/react-native'
// override render method
export { render }
