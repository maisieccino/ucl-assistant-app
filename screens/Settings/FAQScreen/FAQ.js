import PropTypes from "prop-types"
import React from 'react'
import { StyleSheet, View } from "react-native"

import Button from "../../../components/Button"
import Colors from "../../../constants/Colors"
import { Shadow } from "../../../lib"

const styles = StyleSheet.create({
  answerContainer: {
    backgroundColor: Colors.cardBackground,
    borderRadius: 10,
    marginBottom: 30,
    marginTop: 10,
    padding: 20,
    ...Shadow(2),
  },
  question: {
    zIndex: 2,
  },
  section: {

  },
})

class FAQ extends React.Component {
  static propTypes = () => ({
    answer: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.instanceOf(React.Fragment),
    ]),
    question: PropTypes.string,
  })

  constructor() {
    super()
    this.state = {
      showAnswer: false,
    }
  }

  toggleQuestion = () => {
    const { showAnswer } = this.state
    this.setState({ showAnswer: !showAnswer })
  }

  render() {
    const { question, answer } = this.props
    const { showAnswer } = this.state
    return (
      <View style={styles.section}>
        <Button onPress={this.toggleQuestion} style={styles.question}>
          {question}
        </Button>
        {showAnswer && (
          <View style={styles.answerContainer}>
            {answer}
          </View>
        )}
      </View>
    )
  }
}

export default FAQ
