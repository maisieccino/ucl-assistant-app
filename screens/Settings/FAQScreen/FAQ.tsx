import React from 'react'
import { StyleSheet, View } from "react-native"

import Button from "../../../components/Button"
import Colors from "../../../constants/Colors"
import { Shadow } from "../../../lib"

const styles = StyleSheet.create({
  answerContainer: {
    backgroundColor: Colors.cardBackground,
    borderRadius: 10,
    marginBottom: 30,
    marginTop: 10,
    padding: 20,
    ...Shadow(2),
  },
  question: {
    zIndex: 2,
  },
  section: {

  },
})

interface Props {
  answer: string | React.ReactNode,
  question: string,
}

interface State {
  showAnswer: boolean,
}

class FAQ extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props)
    this.state = {
      showAnswer: false,
    }
  }

  toggleQuestion = (): void => {
    const { showAnswer } = this.state
    this.setState({ showAnswer: !showAnswer })
  }

  render(): React.ReactElement {
    const { question, answer } = this.props
    const { showAnswer } = this.state
    return (
      <View style={styles.section}>
        <Button onPress={this.toggleQuestion} style={styles.question}>
          {question}
        </Button>
        {showAnswer && (
          <View style={styles.answerContainer}>
            {answer}
          </View>
        )}
      </View>
    )
  }
}

export default FAQ
