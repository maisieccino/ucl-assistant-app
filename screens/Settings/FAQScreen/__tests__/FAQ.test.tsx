import { fireEvent, render } from '@testing-library/react-native'
import React from 'react'
import { waitForEventLoop } from '../../../../jest/test-utils'
import FAQ from '../FAQ'

describe(`FAQ`, () => {
  const props = {
    answer: `Still a bit stiff`,
    question: `How's the hand?`,
  }

  const setup = () => render(<FAQ {...props} />)

  beforeAll(() => {
    jest.useFakeTimers()
  })

  it(`renders with Q and A`, () => {
    const wrapper = setup()
    expect(wrapper).toMatchSnapshot()
  })

  it(`expands and compresses`, async () => {
    const wrapper = setup()
    const { getByText, queryByText } = wrapper
    const qn = getByText(props.question)

    fireEvent.press(qn)
    expect(wrapper).toMatchSnapshot()

    fireEvent.press(qn)
    await waitForEventLoop()

    expect(queryByText(props.answer)).toBeFalsy()
    expect(wrapper).toMatchSnapshot()
  })
})
