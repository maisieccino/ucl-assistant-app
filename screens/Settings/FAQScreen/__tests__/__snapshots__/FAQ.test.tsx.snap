// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FAQ expands and compresses 1`] = `
<View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "transparent",
        "opacity": 1,
      }
    }
  >
    <ViewManagerAdapter_ExpoLinearGradient
      proxiedProperties={
        Object {
          "colors": Array [
            4279998859,
            4283151357,
          ],
          "endPoint": Array [
            1,
            0,
          ],
          "locations": undefined,
          "startPoint": Array [
            0,
            1,
          ],
        }
      }
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 8,
            "elevation": 2,
            "flexDirection": "row",
            "justifyContent": "center",
            "marginBottom": 5,
            "marginTop": 5,
            "minHeight": 55,
            "minWidth": 150,
            "padding": 8,
            "paddingLeft": 16,
            "paddingRight": 16,
            "shadowColor": "#1B998B",
            "shadowOffset": Object {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.75,
            "shadowRadius": 3,
          },
          Object {
            "zIndex": 2,
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#fefefe",
              "fontFamily": "apercu",
              "fontSize": 20,
              "marginBottom": 1,
              "marginTop": 1,
              "textAlign": "center",
            },
            undefined,
          ]
        }
      >
        How's the hand?
      </Text>
    </ViewManagerAdapter_ExpoLinearGradient>
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#fff",
        "borderRadius": 10,
        "marginBottom": 30,
        "marginTop": 10,
        "padding": 20,
        "shadowColor": "#000",
        "shadowOffset": Object {
          "height": 1,
          "width": 0,
        },
        "shadowOpacity": 0.183,
        "shadowRadius": 1.08,
      }
    }
  >
    Still a bit stiff
  </View>
</View>
`;

exports[`FAQ expands and compresses 2`] = `
<View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "transparent",
        "opacity": 1,
      }
    }
  >
    <ViewManagerAdapter_ExpoLinearGradient
      proxiedProperties={
        Object {
          "colors": Array [
            4279998859,
            4283151357,
          ],
          "endPoint": Array [
            1,
            0,
          ],
          "locations": undefined,
          "startPoint": Array [
            0,
            1,
          ],
        }
      }
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 8,
            "elevation": 2,
            "flexDirection": "row",
            "justifyContent": "center",
            "marginBottom": 5,
            "marginTop": 5,
            "minHeight": 55,
            "minWidth": 150,
            "padding": 8,
            "paddingLeft": 16,
            "paddingRight": 16,
            "shadowColor": "#1B998B",
            "shadowOffset": Object {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.75,
            "shadowRadius": 3,
          },
          Object {
            "zIndex": 2,
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#fefefe",
              "fontFamily": "apercu",
              "fontSize": 20,
              "marginBottom": 1,
              "marginTop": 1,
              "textAlign": "center",
            },
            undefined,
          ]
        }
      >
        How's the hand?
      </Text>
    </ViewManagerAdapter_ExpoLinearGradient>
  </View>
</View>
`;

exports[`FAQ renders with Q and A 1`] = `
<View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "transparent",
        "opacity": 1,
      }
    }
  >
    <ViewManagerAdapter_ExpoLinearGradient
      proxiedProperties={
        Object {
          "colors": Array [
            4279998859,
            4283151357,
          ],
          "endPoint": Array [
            1,
            0,
          ],
          "locations": undefined,
          "startPoint": Array [
            0,
            1,
          ],
        }
      }
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 8,
            "elevation": 2,
            "flexDirection": "row",
            "justifyContent": "center",
            "marginBottom": 5,
            "marginTop": 5,
            "minHeight": 55,
            "minWidth": 150,
            "padding": 8,
            "paddingLeft": 16,
            "paddingRight": 16,
            "shadowColor": "#1B998B",
            "shadowOffset": Object {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.75,
            "shadowRadius": 3,
          },
          Object {
            "zIndex": 2,
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#fefefe",
              "fontFamily": "apercu",
              "fontSize": 20,
              "marginBottom": 1,
              "marginTop": 1,
              "textAlign": "center",
            },
            undefined,
          ]
        }
      >
        How's the hand?
      </Text>
    </ViewManagerAdapter_ExpoLinearGradient>
  </View>
</View>
`;
